src/plugins/intel_cpu/src/infer_request.cpp

```diff
@@ -819,8 +823,12 @@ InferenceEngine::Blob::Ptr ov::intel_cpu::MKLDNNInferRequest::GetBlob(const std:
                     data = make_blob_with_precision(desc);
                     data->allocate();
                 } else {
-                    if (!shape.compatible(ov::PartialShape(data->getTensorDesc().getDims()))) {
-                        IE_THROW(ParameterMismatch) << "Network input and output use the same name: " << name << ", but expect blobs with different shapes.";
+                    const auto& blobDims = data->getTensorDesc().getDims();
+                    if (!shape.compatible(ov::PartialShape(blobDims)) && (!isDynamic ||
+                            std::any_of(blobDims.begin(), blobDims.end(), [](const size_t& dims) { return dims != 0; } ))) {
+                        IE_THROW(ParameterMismatch) << "Network input and output use the same name: " << name
+                                                    << ", but expect blobs with different shapes. Input shape: "
+                                                    << ov::PartialShape(blobDims) << ", output shape: " << shape;
```

Comment:
Good catch! I added check on rank size. It also will work for scalar case, due to fact that scalar can be incompatible only in case of different ranks